cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/export/CgmesExportUtil.java

```diff
@@ -308,6 +309,86 @@ public static <C extends Connectable<C>> void setCgmesTapChangerType(C eq, Strin
+        twt.getLeg1().getOptionalPhaseTapChanger().ifPresent(ptc -> addTapChangerExtension(twt, getTapChangerId(twt, CgmesNames.PHASE_TAP_CHANGER, 1), regulatingControlIsDefined(ptc)));
+
+        twt.getLeg2().getOptionalRatioTapChanger().ifPresent(rtc -> addTapChangerExtension(twt, getTapChangerId(twt, CgmesNames.RATIO_TAP_CHANGER, 2), regulatingControlIsDefined(rtc)));
+        twt.getLeg2().getOptionalPhaseTapChanger().ifPresent(ptc -> addTapChangerExtension(twt, getTapChangerId(twt, CgmesNames.PHASE_TAP_CHANGER, 2), regulatingControlIsDefined(ptc)));
+
+        twt.getLeg3().getOptionalRatioTapChanger().ifPresent(rtc -> addTapChangerExtension(twt, getTapChangerId(twt, CgmesNames.RATIO_TAP_CHANGER, 3), regulatingControlIsDefined(rtc)));
+        twt.getLeg3().getOptionalPhaseTapChanger().ifPresent(ptc -> addTapChangerExtension(twt, getTapChangerId(twt, CgmesNames.PHASE_TAP_CHANGER, 3), regulatingControlIsDefined(ptc)));
+    }
+
+    // If we had alias only for tc1, it will be at end 1
+    // If we had alias for tc1 and tc2, tc2 has been moved to end 1 and combined with tc1, tc1 id will be used
+    // If we only had tc at end 2, it has been moved to end 1 but the id is recorded at end2, tc2 id will be used
+    private static <C extends Connectable<C>> String getTapChangerId(C twt, String cgmesTapChangerTag) {
+        String aliasType1 = Conversion.CGMES_PREFIX_ALIAS_PROPERTIES + cgmesTapChangerTag + 1;
+        Optional<String> optionalTapChangerId1 = twt.getAliasFromType(aliasType1);
+        if (optionalTapChangerId1.isPresent()) {
+            return optionalTapChangerId1.get();
+        } else {
+            String aliasType2 = Conversion.CGMES_PREFIX_ALIAS_PROPERTIES + cgmesTapChangerTag + 2;
+            Optional<String> optionalTapChangerId2 = twt.getAliasFromType(aliasType2);
+            if (optionalTapChangerId2.isEmpty()) {
+                // We create a new id always at end 1
+                String newTapChangerId = CgmesExportUtil.getUniqueId();
+                twt.addAlias(newTapChangerId, aliasType1);
+                return newTapChangerId;
+            } else {
+                return optionalTapChangerId2.get();
+            }
+        }
+    }
+
+    private static <C extends Connectable<C>> String getTapChangerId(C twt, String cgmesTapChangerTag, int endNumber) {
+        String aliasType = Conversion.CGMES_PREFIX_ALIAS_PROPERTIES + cgmesTapChangerTag + endNumber;
+        Optional<String> optionalTapChangerId = twt.getAliasFromType(aliasType);
+        if (optionalTapChangerId.isEmpty()) {
+            String newTapChangerId = CgmesExportUtil.getUniqueId();
+            twt.addAlias(newTapChangerId, aliasType);
+            return newTapChangerId;
+        } else {
+            return optionalTapChangerId.get();
+        }
+    }
+
+    private static boolean regulatingControlIsDefined(RatioTapChanger rtc) {
+        return !Double.isNaN(rtc.getTargetV()) && rtc.getTargetV() > 0.0;
```

Comment:
we should also check that regulating terminal is not null

Reply:
and/or verify the flag for load tap changing capabilities

Reply:
Done. Only the regulating terminal checking is needed